Restructured getgroups check out of PIE

angieenee opened the pull request:
Previously, changes made would check the necessity of calling setgroups
in restorer.c of pie. This commit relocates this check to criu/restore.c
before restore_creds occurs.

Signed-off-by: Angie Ni <avtni@google.com>

Commit:
Restructured getgroups check out of PIE

Previously, changes made would check the necessity of calling setgroups
in restorer.c of pie. This commit relocates this check to criu/restore.c
before restore_creds occurs.

Signed-off-by: Angie Ni <avtni@google.com>

angieenee commented on criu/cr-restore.c:
```diff
@@ -3235,7 +3238,18 @@ rst_prep_creds_args(CredsEntry *ce, unsigned long *prev_pos)
+	n = getgroups(0, NULL);
+	len = n * sizeof(unsigned int);
+	gids = (unsigned int*)xmalloc(len * sizeof(unsigned int));
+	if (gids == MAP_FAILED) {
```
xmalloc will never return MAP_FAILED -- check xmalloc return values

kashyapp commented on criu/cr-restore.c:
```diff
@@ -3133,6 +3133,9 @@ rst_prep_creds_args(CredsEntry *ce, unsigned long *prev_pos)
 	struct thread_creds_args *args;
+	int n, len;
+	bool groups_are_ok;
+	unsigned int* gids;
```
Shouldn't we  use `gid_t` instead of `unsigned int`

angieenee replied:
Ah, I think that's right. We memcmp gids to ce->groups. Both `gid_t` and ce->groups are `uint32_t`.

kashyapp commented on criu/cr-restore.c:
```diff
@@ -3133,6 +3133,9 @@ rst_prep_creds_args(CredsEntry *ce, unsigned long *prev_pos)
 {
 	unsigned long this_pos;
 	struct thread_creds_args *args;
+	int n, len;
```
`n` and `len` are much too far away from point of use to be named so simply.

kashyapp replied:
Why are they declared so far away though? Is it because one of the previous PR comments was to declare at the top of the function? This very function has a number of variables declared in a number of inner scopes.

mclapinski replied:
Yeah, but you need a scope to declare it in. The variables has to be at the top of the scope and that is the top of the scope in which they are used.

angieenee replied:
I can rename them to accommodate for the distance.

mclapinski replied:
No need if you move the code to a separate function

kashyapp commented on criu/cr-restore.c:
```diff
@@ -3235,7 +3238,18 @@ rst_prep_creds_args(CredsEntry *ce, unsigned long *prev_pos)
 	memcpy(args->cap_bnd, ce->cap_bnd, sizeof(args->cap_bnd));
 
-	if (ce->n_groups) {
+	n = getgroups(0, NULL);
```
@mclapinski should we extract this logic into a standalone function?

mclapinski replied:
Probably yes. This function is long enough.

kashyapp commented on criu/cr-restore.c:
```diff
@@ -3235,7 +3238,18 @@ rst_prep_creds_args(CredsEntry *ce, unsigned long *prev_pos)
 	memcpy(args->cap_bnd, ce->cap_bnd, sizeof(args->cap_bnd));
 
-	if (ce->n_groups) {
+	n = getgroups(0, NULL);
```
Btw, Andrei had another comment that I think we haven't addressed yet. 
"you don't need to call sys_getgroups if n isn't equal to ce->n_groups"

angieenee commented on criu/cr-restore.c:
```diff
@@ -3128,6 +3128,22 @@ static void rst_reloc_creds(struct thread_restore_args *thread_args,
 	thread_args->creds_args = args;
 }
 
+bool groups_are_okay(CredsEntry *ce) {
```
Make this static bool

mclapinski commented on criu/cr-restore.c:
```diff
@@ -3128,6 +3128,22 @@ static void rst_reloc_creds(struct thread_restore_args *thread_args,
 	thread_args->creds_args = args;
 }
 
+static bool groups_are_okay(CredsEntry *ce) {
```
Pass only the parts of CredsEntry that you use

mclapinski commented on criu/cr-restore.c:
```diff
@@ -3128,6 +3128,22 @@ static void rst_reloc_creds(struct thread_restore_args *thread_args,
+	check_num_groups = getgroups(0, NULL);
+	if (check_num_groups != ce->n_groups)
+		return false;
+	gids_len = check_num_groups * sizeof(unsigned int);
```
Before this line check if check_num_groups is 0.

kashyapp commented on criu/cr-restore.c:
```diff
@@ -3128,6 +3128,24 @@ static void rst_reloc_creds(struct thread_restore_args *thread_args,
+	if (check_num_groups == 0)
+		return true;
+	gids_len = check_num_groups * sizeof(unsigned int);
+	gids = (gid_t*)xmalloc(gids_len * sizeof(gid_t));
```
why is there a `* sizeof(gid_t)` again here? gids_len should already be what is needed.

kashyapp replied:
Because of this twice calculation, we're doign a `malloc(n * sizeof(unsigned int) * sizeof(gid_t))`

angieenee replied:
Nice catch! I think an earlier change in variable types caused this issue.

kashyapp commented on criu/cr-restore.c:
```diff
@@ -3128,6 +3128,24 @@ static void rst_reloc_creds(struct thread_restore_args *thread_args,
 }
 
+static bool groups_are_okay(gid_t* groups, int n_groups) {
+	int check_num_groups, gids_len;
```
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#naming

Given that you've extracted into a function, I do think going back to `n, len` is a good idea.

kashyapp commented on criu/cr-restore.c:
```diff
@@ -3128,6 +3128,24 @@ static void rst_reloc_creds(struct thread_restore_args *thread_args,
 
+static bool groups_are_okay(gid_t* groups, int n_groups) {
+	int check_num_groups, gids_len;
+	bool groups_are_ok;
```
likewise, this can be `ret`

Commit:
updated groups_are_okay variable names, xmalloc

Previously, groups_are_okay had specific variable names because the
helpef function was initially used in rst_prep_creds_args. I was also
xmallocing quadratic memory by using sizeof() twice.

Signed-off-by: Angie Ni <avtni@google.com>

mclapinski reviewed (approved):
nice work, remember to squash
